### PR TITLE
test.py: test cases

### PR DESCRIPTION
**Pull request by alecco**
Handle cases per test from command line and add internal logic. Allow splitting pytest test by case.

Major speedup by only loading suites and tests specified in command line (if specified). Before this PR, all suites would be loaded, including `boost` suite running all its 168 binaries to collect the cases.

Syntax changed to `suite::test::case` with possible patterns  `suite`, `suite::test`, `suite::test::case`, `::test`, `::test::case`.

Previous PR discussion for pytest case: https://github.com/scylladb/scylladb/pull/13950 (closed).

### CODE REVIEW

**Comment by alecco** [comment]
Currently the PR splits all tests into individual cases. I plan to add not splitting cases for `dev` and `release` by default.

And later we could un-split some test cases.

**Comment by alecco** [comment]
Changes in https://github.com/scylladb/scylladb/commit/3df973341b0a6e6a395b442e2a8cc99dc7858755:

- fix test pattern filter

File: test.py
Comment (nyh) [inline test.py]: Note: Pytest usually uses "filename.py" before the "::". I think (but I'm not sure) your self.name drops the ".py". It's not terrible, but we should be aware of this difference.
Also, despite my recommendation, it seems some suites used classes inside the test files, so the test path gains yet another component - the class name.  I guess this isn't supported here? (I'm not too sad, I think these test classes are a mistake anyway).

File: test.py
Comment (nyh) [inline test.py:488]: Why do you need to list the cases, if anyway the UI is to run one specific case and not some glob like "all tests beginning with the letter x"?

File: test.py
Comment (nyh) [inline test.py:967]: Why "-k" and not just give the file name and case name?
I'm worried that it can pick up similarly-named tests from other files as well, and other strange things (I don't know exactly what "-k" does...).

**Review by nyh — COMMENTED** [review]
I like this direction! I'm always running individual test files and individual test functions through test/*/run, so this will give this ability also to people who use test.py.
But I'm not familar enough with the test.py code to thoroughly review it. I just left a few comments.

File: test.py
Comment (alecco) [reply test.py]: Makes sense.

Comment (nyh) [reply test.py]: By the way, some of the source files have a path, e.g., cql-pytest's `cassandra_tests/validation/entities/type_test.py`. Is this full path in self.name (it should...)?

File: test.py
Comment (kbr-scylla) [inline test.py:278]: This commit is mixing logical changes with stylistic changes including indentation, making it much harder to review than it could have been.

For example, I see that `add_test_tasks` is basically the same thing as previously `pending`.

And instead of
```
if len(pending) == 0:
    return
... logic ...
```

you have:
```
if add_test_tasks:
    ... logic ...
```

where `... logic ...` is exactly the same, just indented differently.

Such stylistic changes should really have their own commits.

File: test.py
Comment (kbr-scylla) [inline test.py:345]: Another case of mixing stylistic changes and code movement with logical changes, making job harder than necessary for reviewers.

Comment (kbr-scylla) [inline test.py:345]: Please add a docstring to this function. Describe what `exe` is and what the function does.
Does `exe` in the function name start for `execute`? Please don't abbreviate this

Comment (kbr-scylla) [reply test.py:345]: Please split the different types of changes into separate commits.

**Review by kbr-scylla — COMMENTED** [review]
> Handle cases per test from command line and add internal logic.

This is too vague, please be more precise and provide an example.

File: test.py
Comment (alecco) [reply test.py:251]: test file, yes. I'll clarify.

File: test.py
Comment (alecco) [reply test.py:262]: > Does `run_first_tests` use the new test name format?

Good point. Yes, `run_first_tests` should have this format and I missed to update the configs. Will do.

> I think that converting to `str` in order to sort is just asking for silent bugs that we never catch.

Good catch, the second `str(x)` should be just `x`.

> Instead we should go for strong types and turn run_first_tests into a set of TestCases. Parse when constructing run_first_tests and validate format correctness.

Makes sense.

File: test.py
Comment (kostja) [reply test.py:211]: The reason shortname came to exist initially was that the test name could contain, for example, test extension, such as .py or _test.cql, and its shortname wouldn't. Besides, the test name doesn't contain suite name, and it was a contrast with a fully qualified test name, such as test_suite.test_file.

I think if we want to achieve some sort of clarity with all these names, we need to stick to these terms:
- test suite name
- test file name
- test case name

So, right now, shortname is a short name of the test file. It came from the times when 1 file = 1 test. Now this variable name is ambiguous I agree, but "name" is even more ambiguous. Perhap test_file_short or short_test_file_name, or something would be less accurate.

File: test.py
Comment (kostja) [reply test.py:39]: @alecco I miss the motivation for changing the name separator. :: against . is 2-4x less efficient, depending how you count (characters or dots). Dots are really easier to type and easier to remember. Why did you decide to replace the separator?

File: test.py
Comment (kostja) [inline test.py]: And here the ambiguity from the previous rename is striking us. Suite_name is fine, name should be "file_name", and "case" should be "case_name". Or, if you like brevity, suite, file, case.

File: test.py
Comment (alecco) [reply test.py:39]: It's more standard and pytest (and many other frameworks) use `test::case` (as @nyh pointed out).

I added `suite::test::name` for simplicity, because `suite.name::case` looks ugly IMO.

File: test.py
Comment (alecco) [reply test.py:211]: I plan to rename `TestCase` to `TestDef` (suggestions welcome). So it would be `TestDef.name`, better than `TestDef.test` IMO.

File: test.py
Comment (kostja) [reply test.py]: I don't see what pytest has to do with boost or unit tests, so we can't support what pytest supports anyway. cql tests don't have test cases, at least not now.

File: test.py
Comment (alecco) [reply test.py]: OK, `file_name` and `case_name`.

File: test.py
Comment (nyh) [reply test.py:39]: Besides being used by pytest (which is the basis for most of our test frameworks - dtest, topology, cql-pytest, alternator, etc.), there also a bizarre fact (in my experience - please check again in case I made a mistake) that when you write in a **github** issue `suite::somefilename::sometestname`  it allows searching each individual component (e.g., "sometestname") which is useful when looking for duplicate issues, while if you write `suite.somefilename.sometestname` the components aren't searchable (at least in my experience).

Another issue is that at least in pytest, the test file name does include the extension (e.g., .py), so a dot can't be used as a separator.

File: test.py
Comment (alecco) [reply test.py]: In my [case PR](https://github.com/scylladb/scylladb/pull/15238)  when there's no case it's omitted like `suite::test_file`.

Comment (nyh) [reply test.py]: @kostja I don't understand your question. Let me try to explain, hoping it's related to your question:
test.py can run different types of tests - pytest tests, boost tests, cql approval tests, and so on. So we're looking for some uniform syntax for representing the name of a test to run (whole suite, whole file, or just function, whenever that's possible) that can be inspired by the syntax used by one of the above frameworks (I propose pytest) but needs to be translated to each.

So for example `cql-pytest::test_uda.py::test_custom_avg` can refer to one test function in a cql-pytest, `boost::wasm_test::test_long_udf_yields` can refer to one test function in Boost, and so on.

My comment above was that the test-file part of a cql-pytest can include the ".py" (I guess we can make it optional) and also be a pathname with slashes in it, like `cassandra_tests/validation/entities/type_test.py`. Elsewhere I also noted that in dtest (but I think in none of the scylla.git frameworks?) pytest tests can also be inside classes, you you have yet another component between the file name and test function name - the class name.

File: test.py
Comment (kostja) [reply test.py:39]: I think we should use suite.file.test, or suite/file/test, since this is less typing on an US-ASCII keyboard layout. But OK, I see I am outvoted. Let's make it :: if such is the popular consensu. I believe, however, in pytest, weirdly, it's both dots and semicolons, so at least let's please not do it, let's consistently use semicolons...

File: test.py
Comment (kostja) [reply test.py:488]: The current semantics is to actually run all matches, so that you can type ./test.py raft and that would run *everything* in *all suites* that has name "raft" in it.

Comment (alecco) [reply test.py:488]: What would you think of `./test.py raft` for matching any and `./test.py raft::` for matching the suite raft? We need to make it unambiguous.

Comment (kostja) [reply test.py:488]: Exactly. It is working like this today. The match is made against a fully qualified test name, i.e. suite_name.test_file_name. After your change, it will be against a fully qualified test name as well, but the syntax will be suite_name::test_file_name::test_case_name. But it's still going to be pattern match, not exact match, so it'll be fine.

**Comment by alecco** [comment]
Sent an update before doing renames, so reviewers can check the diff easily. But more updates will come in subsequent versions.

Changes in https://github.com/scylladb/scylladb/commit/38de3c067b5f6e2bce99011c55710ca92ffc4968 ([diff to prev](https://github.com/scylladb/scylladb/compare/b470b7e8b0ec874c3ff0ec627916150cd6fb575e..38de3c067b5f6e2bce99011c55710ca92ffc4968)):

- `TestCase` docstring (Kamil)
- Fix `TestCase` representation to handle no test (test_file) specified
- `_test_defs()` docstring (Kamil?)
- `_test_defs()` sort by tuple not str(x) (Kamil)
- Use `.py` extension in test name for `PythonTest` (Nadav)
- Fix `Test.uname` (used for logs) for the case of no case specified (was creating `.*.1.log`) 
- Fix `test_path_of_elements()` to handle no case specified
- Pass `test_file.py::case` instead of `-k test_file.py::case` to pytest (Nadav)

File: test.py
Comment (alecco) [reply test.py]: Fixed in https://github.com/scylladb/scylladb/commit/38de3c067b5f6e2bce99011c55710ca92ffc4968

File: test.py
Comment (alecco) [reply test.py]: Fixed in https://github.com/scylladb/scylladb/commit/38de3c067b5f6e2bce99011c55710ca92ffc4968  (will change to `TestDef.test_file` in subsequent update.

File: test.py
Comment (kostja) [reply test.py:230]: Does TestDef stand for a test case or a testfile?

Comment (alecco) [reply test.py:230]: > Does TestDef stand for a test case or a testfile?

One test to run. Either with a case or the whole file.

And also test pattern to match (from command line). Maybe I should split this to its own class.

About the name, I'm open to suggestions, as I stated above. Or we can rename it later.

File: test.py
Comment (nyh) [reply test.py:39]: "suite/file/test" is problematic when the "file" part can have subdirectories (as we have in cql-pytest, and I see no reason not to have in other directories as well).

What do you mean that pytest has "both dots and semicolons"? Where do you use dots in pytest?

File: test.py
Comment (kostja) [reply test.py]: So @alecco the commit message contains a new term, "test instance", you also use "TestDef" elsewhere in our discussion, is "Def" what "defines" a test instance?

Comment (kostja) [reply test.py]: Perhaps "Test instance" is a "TestTask", and "TestDef" is a "TestTaskDef". In a comment for TestTask we could clarify that a TestTask can stand for an entire test file or a single test case. Alternatively we could say that we always run an entire test file and test case selection is a filter that we push down to the runner to filter out unnecessary cases. In this case the terminology would be TestFileTask and TestFileTaskDef or something along these lines. WDYT?

Comment (alecco) [reply test.py]: The (current) docstring for `class TestCase`'s `.match()` method compares 2 object instances of TestCase. The this/self instance and the other one passed in the parameters. This is Python terminology.

File: test.py
Comment (kostja) [reply test.py:93]: Please follow up with the naming convention we've established down to this commit (name -> test_file_name)

File: test.py
Comment (kostja) [inline test.py:1251]: I was hoping wildcards would continue being supported. ./test.py raft is a convenient handle for running everything related to raft.

File: test.py
Comment (kostja) [inline test.py:1391]: Overall I like the idea of saving on the load/preprocess time if wildcard is specified very match. But I think it kills the idea of a general purpose pattern that can be specified for 'names'. E.g. if I type ./test.py raft:: is it selecting a single suite or a single test file with a matching name in all suites? It would be lovely to have a powerful filter capabilities and fast search for the matching tests, something I can't think of a way of combining in a single pattern match.

File: test.py
Comment (kostja) [inline test.py]: Adds extra painful typing :/ and makes it hard to remember. @nyh @kbr-scylla @denesb do you really think that compatibility with dtest/pytest should lead to so much extra typing?

File: test.py
Comment (kostja) [inline test.py:146]: it would be lovely to see a more elaborate rationale for this change.

File: test.py
Comment (nyh) [reply test.py]: I think the ".py" can be optional - but if someone types it (because they are familiar with pytest, perhaps), what does it hurt to allow it as well?

File: test.py
Comment (alecco) [reply test.py]: > Perhaps "Test instance" is a "TestTask", and "TestDef" is a "TestTaskDef". In a comment for TestTask we could clarify that a TestTask can stand for an entire test file or a single test case. Alternatively we could say that we always run an entire test file and test case selection is a filter that we push down to the runner to filter out unnecessary cases. In this case the terminology would be TestFileTask and TestFileTaskDef or something along these lines. WDYT?

TestFileTask` is OK. What could be the name for the pattern matching class? (built from the command line "name" specification).

File: test.py
Comment (nyh) [reply test.py:1251]: Can you give an example of when such wildcards are useful to you? In my experience I always noticed the exact opposite - you type some word "xyz" and you get a few unrelated tests that happen to contain this word in the middle of the name.
That being said, we could always fall back to allowing a wildcard as well - if the test path doesn't have a "::" and isn't one of the small number of suite names - it is looked up as a substring search.

File: test.py
Comment (alecco) [reply test.py:1391]: I wrote in another review comment:

>>> What would you think of `./test.py raft` for matching any and `./test.py raft::` for matching the suite raft? We need to make it unambiguous.

**Comment by alecco** [comment]
Changes in https://github.com/scylladb/scylladb/commit/180b5d6eff44ce0033a5e8e060bb468eef2d86e6:

- [didn't send the renames yet because it would change everything, will send later]
- change `suite.yaml` `no_parallel_cases` to `parallel_cases` so tests needing parallel run of cases are specified (before all boost test files were parallelized except for 2); only parallelize boost test files taking more than a minute (Avi)
- ditched per mode parallel spec as ^^ fixes the problem and makes it better suited overall (but not perfect)
- new `TestCase.startswith()` for matching
- `run_first` now supports case names (Kamil)

Changes in https://github.com/scylladb/scylladb/commit/2904b235f35f2dd999f5b294c5b45f31ec00edfb:

- rebased on latest master

**Comment by alecco** [comment]
Changes in https://github.com/scylladb/scylladb/commit/2e18e83ea72b014ccb93d878609a0cc8ef2430b5:

- Rebased on latest https://github.com/scylladb/scylladb/pull/15519
- fixed match patterns for disabled and skip
- Removed `Test.case_name` and just use `Test.test_def.case_name`
- Fixed `PythonTest` test_file_name and pytest argument path with suite
- fix `print_summary()` using old `Test.name`
- added type to `print_summary()` `failed_tests` (this was hiding ^^)

**Comment by alecco** [comment]
Changes in https://github.com/scylladb/scylladb/commit/c7b3baf27771af8c84d318af470dfacba67893e0:

- Mark disable in config and not with defines (fixes debug mode `boost::cache_algorithm_test` issue)
- configure slow raft tests for parallel execution

**Comment by benipeled** [comment]
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3947/

CI ran on spider7 and lost connection, I can see `BUG: kernel NULL pointer...`
@alecco any chance something in your change causing a memory leak or other infra issues? 

```
Sep 22 16:13:53 spider7.cloudius-systems.com sssd_be[1222]: GSSAPI client step 1
Sep 22 16:13:53 spider7.cloudius-systems.com sssd_be[1222]: GSSAPI client step 2
Sep 22 16:21:49 spider7.cloudius-systems.com kernel: __vm_enough_memory: 17 callbacks suppressed
Sep 22 16:21:49 spider7.cloudius-systems.com kernel: __vm_enough_memory: pid: 1312089, comm: alternator_unit, not enough memory for the allocation
Sep 22 16:21:49 spider7.cloudius-systems.com kernel: __vm_enough_memory: pid: 1312089, comm: alternator_unit, not enough memory for the allocation
Sep 22 16:21:49 spider7.cloudius-systems.com kernel: __vm_enough_memory: pid: 1312089, comm: alternator_unit, not enough memory for the allocation
Sep 22 16:21:49 spider7.cloudius-systems.com kernel: __vm_enough_memory: pid: 1312089, comm: alternator_unit, not enough memory for the allocation
Sep 22 16:21:49 spider7.cloudius-systems.com kernel: __vm_enough_memory: pid: 1312087, comm: alternator_unit, not enough memory for the allocation
Sep 22 16:21:49 spider7.cloudius-systems.com kernel: __vm_enough_memory: pid: 1312087, comm: alternator_unit, not enough memory for the allocation
Sep 22 16:21:49 spider7.cloudius-systems.com kernel: __vm_enough_memory: pid: 1312089, comm: alternator_unit, not enough memory for the allocation
Sep 22 16:21:49 spider7.cloudius-systems.com kernel: __vm_enough_memory: pid: 1312089, comm: alternator_unit, not enough memory for the allocation
Sep 22 16:21:49 spider7.cloudius-systems.com kernel: __vm_enough_memory: pid: 1312087, comm: alternator_unit, not enough memory for the allocation
Sep 22 16:21:49 spider7.cloudius-systems.com kernel: __vm_enough_memory: pid: 1312089, comm: alternator_unit, not enough memory for the allocation
Sep 22 16:24:50 spider7.cloudius-systems.com kernel: BUG: kernel NULL pointer dereference, address: 0000000000000062
Sep 22 16:24:50 spider7.cloudius-systems.com kernel: #PF: supervisor read access in kernel mode
Sep 22 16:24:50 spider7.cloudius-systems.com kernel: #PF: error_code(0x0000) - not-present page
Sep 22 16:24:50 spider7.cloudius-systems.com kernel: PGD 1d2492067 P4D 1d2492067 PUD 344473067 PMD 0 
Sep 22 16:24:50 spider7.cloudius-systems.com kernel: Oops: 0000 [#1] PREEMPT SMP NOPTI
Sep 22 16:24:50 spider7.cloudius-systems.com kernel: CPU: 5 PID: 1325258 Comm: tracing_test Tainted: G        W          6.4.15-100.fc37.x86_64 #1
Sep 22 16:24:50 spider7.cloudius-systems.com kernel: Hardware name: ASRock B650 PG Lightning/B650 PG Lightning, BIOS 1.24.AS02 05/02/2023
Sep 22 16:24:50 spider7.cloudius-systems.com kernel: RIP: 0010:next_uptodate_page+0x45/0x1f0
Sep 22 16:24:50 spider7.cloudius-systems.com kernel: Code: 0f 84 2f 01 00 00 48 81 ff 06 04 00 00 0f 84 a3 00 00 00 48 81 ff 02 04 00 00 0f 84 26 01 00 00 40 f6 c7 01 0f 85 8c 00 00 00 <48> 8b 07 a8 01 0f 85 81 00 00 00 8b 47 34 85 c0 74 7a 8d 50 01 4c
Sep 22 16:24:50 spider7.cloudius-systems.com kernel: RSP: 0000:ffffa9251a44fce8 EFLAGS: 00010246
Sep 22 16:24:50 spider7.cloudius-systems.com kernel: RAX: 0000000000000062 RBX: ffffa9251a44fd38 RCX: 00000000000045ee
Sep 22 16:24:50 spider7.cloudius-systems.com kernel: RDX: ffffa9251a44fd38 RSI: ffff89941f30e2b0 RDI: 0000000000000062
Sep 22 16:24:50 spider7.cloudius-systems.com kernel: RBP: ffff89941f30e2b0 R08: 0000000000fc1000 R09: 0000000004b73fff
Sep 22 16:24:50 spider7.cloudius-systems.com kernel: R10: ffff899ea0c01c0c R11: 0000000000000000 R12: 00000000000045ee
Sep 22 16:24:50 spider7.cloudius-systems.com kernel: R13: ffff898f0a9d26e0 R14: ffff89941f30e2b0 R15: ffff898fd2492000
Sep 22 16:24:50 spider7.cloudius-systems.com kernel: FS:  00007f15adbce000(0000) GS:ffff89add8340000(0000) knlGS:0000000000000000
Sep 22 16:24:50 spider7.cloudius-systems.com kernel: CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
Sep 22 16:24:50 spider7.cloudius-systems.com kernel: CR2: 0000000000000062 CR3: 0000000d23a6a000 CR4: 0000000000750ee0
Sep 22 16:24:50 spider7.cloudius-systems.com kernel: PKRU: 55555554
Sep 22 16:24:50 spider7.cloudius-systems.com kernel: Call Trace:
```

**Comment by nyh** [comment]
What's holding this PR?

I'm constantly annoyed why developing a single test in a larger test file that it needs to re-run all the tests in the file, and really want this feature to go in.

Please make sure that the commit message is up-to-date on what you ended up doing (colons or dots or both or what?) . Please also update docs/dev/testing.md to explain how to run a single test function. Please also mention whether it works for all suites (also Boost? It could!) or just Python-based suites.

File: test.py
Comment (nyh) [inline test.py:1500]: "Improves 40 to 20 seconds"?!

```
time test/cql-pytest/run test_json.py::test_fromjson_timestamp_submilli
```
Currently takes 1 second (0.3 seconds to boot Scylla, 0.17 seconds to run the test, and about half a second more for I am not sure exactly what).

Are you saying that if I use test.py to run a single test it will take 20 seconds (at least not 40 seconds!) just to figure out what I meant?

I think that you must have a special case for running a single test case or a single test file. This special case should NOT try to discover all other tests.

File: test.py
Comment (nyh) [inline test.py:527]: I am not against adding ".py" to test file names in the command (this is what pytest does so I'm used to it), but please make sure you update the examples in the documentation (doc/dev/testing.md) and commit messages to include the ".py" that is now needed (if I understand correctly).

File: test/cql-pytest/suite.yaml
Comment (nyh) [inline test/cql-pytest/suite.yaml:20]: I just tested randomly, and this entire test file takes 0.7 seconds on my laptop. What made you decide it needs to be parallelized internally?

**Review by nyh — APPROVED** [review]
I like this direction, and although I don't understand all the details (and left a few questions and comments), I would like this to go in and then we can continue to improve it incrementally.
However, we'll need also the approval of one of the bigger test.py experts (Kostja or Kamil).

**Comment by kbr-scylla** [comment]
@nyh Alejo is not working for Scylla anymore, the PR is dead.

**Comment by nyh** [comment]
> @nyh Alejo is not working for Scylla anymore, the PR is dead.

1. I missed that. Thanks.
2. I don't think the first part of the sentence implies the second part. It may (if you disapprove of this PR, and you are the person who most cares about this code). But it also may not. The Scylla source code is full of code written by people who don't work in Scylla any more.

**Pull request closed without merging**
